File: guild/src/main/java/com/os/operando/guild/Octet.java
package com.os.operando.guild;

/**
 * A tuple of eight elements.
 *
 * @param <F>  first element type
 * @param <S>  second element type
 * @param <T>  third element type
 * @param <FO> fourth element type
 * @param <FI> five element type
 * @param <SI> six element type
 * @param <SE> seven element type
 * @param <E>  eight element type
 */
public class Octet<F, S, T, FO, FI, SI, SE, E> {

    private final F first;
    private final S second;
    private final T third;
    private final FO fourth;
    private final FI five;
    private final SI six;
    private final SE seven;
    private final E eight;

    private Octet(final F first, final S second, final T third, final FO fourth, final FI five, final SI six, final SE seven, final E eight) {
        this.first = first;
        this.second = second;
        this.third = third;
        this.fourth = fourth;
        this.five = five;
        this.six = six;
        this.seven = seven;
        this.eight = eight;
    }

    public static <F, S, T, FO, FI, SI, SE, E> Octet<F, S, T, FO, FI, SI, SE, E> create(F first, S second, T third, FO fourth, FI five, SI six, SE seven, E eight) {
        return new Octet<>(first, second, third, fourth, five, six, seven, eight);
    }

    public F getFirst() {
        return first;
    }

    public S getSecond() {
        return second;
    }

    public T getThird() {
        return third;
    }

    public FO getFourth() {
        return fourth;
    }

    public FI getFive() {
        return five;
    }

    public SI getSix() {
        return six;
    }

    public SE getSeven() {
        return seven;
    }

    public E getEight() {
        return eight;
    }
}

File: guild-kotlin/src/main/kotlin/com/os/operando/guild/Octet.kt
package com.os.operando.guild

/**
 * A tuple of eight elements.
 *
 * @param F  first element type
 * @param S  second element type
 * @param T  third element type
 * @param FO fourth element type
 * @param FI five element type
 * @param SI six element type
 * @param SE seven element type
 * @param E eight element type
 * @property first First value
 * @property second Second value
 * @property third Third value
 * @property fourth Fourth value
 * @property five Five value
 * @property six Six value
 * @property seven Seven value
 * @property eight Eight value
 */
data class Octet<out F, out S, out T, out FO, out FI, out SI, out SE, out E>(
        val first: F,
        val second: S,
        val third: T,
        val fourth: FO,
        val five: FI,
        val six: SI,
        val seven: SE,
        val eight: E) {
}

infix fun <A, B, C, D, E, F, G, H, I> Octet<A, B, C, D, E, F, G, H>.to(that: I): Ennead<A, B, C, D, E, F, G, H, I>
        = Ennead(this.first, this.second, this.third, this.fourth, this.five, this.six, this.seven, this.eight, that)